### DISCUSSION TITLE
Can IService Collection inject all variants of generic types?

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/runtime#105095] [Q&A]
Similar to XmlSerializer(typeof(T)) and ArrayPool<T>

**Comment by Joe4evr** [comment]
You can do `services.Add{Lifetime}(typeof(ISomeService<>), typeof(Impl<>))` (or `services.Add{Lifetime}(typeof(Impl<>))`).

You can also register specializations in addition to the open service: `services.Add{Lifetime}<ISomeService<Foo>, FooService>()`
* Do note: I don't know if this is a bug or not, but it seems this only works if the specialization is registered *after* the open service

**Reply by zms9110750** [reply]
Can I specify the factory and key?
I want to do:
```csharp
build.Services.AddSingleton(typeof(IMemoryOwner<>), sp => MemoryPool<>.Shared.Rent());
build.Services.AddKeyedSingleton(serviceType: typeof(XmlSerializer), serviceKey: { all }, (sp, key) => new XmlSerializer((Type)key));
```
